File: 0637-average-of-levels-in-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        if (root == nullptr) {
            return ans;
        }

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty())
        {
            int64_t  sum = 0;
            int size = q.size();

            for (size_t i = 0; i < size; i++)
            {
                TreeNode* node = q.front();
                q.pop();

                sum = sum + node->val;

                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }

            double avg = static_cast<double>(sum) / static_cast<double>(size);
            ans.push_back(avg);
        }


        return ans; 

    }
};

File: 0637-average-of-levels-in-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
	ans := []float64{}
	if root == nil {
		return ans
	}

	q := []*TreeNode{}
    q = append(q, root)

	for len(q) > 0 {
		sum := 0
		size := len(q)

		for i := 0; i < size; i++ {
			node := q[0]
			q = q[1:]

			sum += node.Val
			
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}

		avg := float64(sum) / float64(size)
		ans = append(ans, avg)

	}

	return ans
}
